File: flutter_app/lib/models/driver.dart
class DriverModel {
  final String id;
  final String name;
  final String phone;
  final String? loginEmail;
  final List<DriverLeaveModel> leaves;
  final double salaryPerDay;
  final double salaryPerTrip;
  final double bataRate;
  final double totalWorkingHours;
  final int totalWorkingDays;
  final int totalTripsCompleted;
  final double totalBataEarned;

  const DriverModel({
    required this.id,
    required this.name,
    required this.phone,
    this.loginEmail,
    this.leaves = const [],
    this.salaryPerDay = 0,
    this.salaryPerTrip = 0,
    this.bataRate = 0,
    required this.totalWorkingHours,
    required this.totalWorkingDays,
    this.totalTripsCompleted = 0,
    this.totalBataEarned = 0,
  });

  factory DriverModel.fromJson(Map<String, dynamic> json) {
    final user = json['userId'];

    return DriverModel(
      id: json['_id'] as String,
      name: json['name'] as String,
      phone: json['phone'] as String,
      loginEmail: user is Map<String, dynamic> ? (user['email'] as String?) : null,
      leaves: (json['leaves'] as List<dynamic>? ?? const [])
          .whereType<Map<String, dynamic>>()
          .map(DriverLeaveModel.fromJson)
          .toList(),
        salaryPerDay: (json['salaryPerDay'] as num?)?.toDouble() ?? 0,
        salaryPerTrip: (json['salaryPerTrip'] as num?)?.toDouble() ?? 0,
        bataRate: (json['bataRate'] as num?)?.toDouble() ?? 0,
      totalWorkingHours: (json['totalWorkingHours'] as num?)?.toDouble() ?? 0,
      totalWorkingDays: (json['totalWorkingDays'] as num?)?.toInt() ?? 0,
      totalTripsCompleted: (json['totalTripsCompleted'] as num?)?.toInt() ?? 0,
      totalBataEarned: (json['totalBataEarned'] as num?)?.toDouble() ?? 0,
    );
  }
}

class DriverLeaveModel {
  final String id;
  final DateTime from;
  final DateTime to;
  final String reason;
  final String status;

  const DriverLeaveModel({
    required this.id,
    required this.from,
    required this.to,
    required this.reason,
    required this.status,
  });

  factory DriverLeaveModel.fromJson(Map<String, dynamic> json) {
    return DriverLeaveModel(
      id: json['_id'] as String,
      from: DateTime.parse(json['from'] as String),
      to: DateTime.parse(json['to'] as String),
      reason: (json['reason'] as String?) ?? '-',
      status: (json['status'] as String?) ?? 'pending',
    );
  }
}

File: backend/models/Driver.js
const mongoose = require('mongoose');

const leaveSchema = new mongoose.Schema(
  {
    from: { type: Date, required: true },
    to: { type: Date, required: true },
    reason: { type: String, required: true },
    status: { type: String, enum: ['pending', 'approved', 'rejected'], default: 'pending' },
    approvedBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },
  },
  { timestamps: true }
);

const driverSchema = new mongoose.Schema(
  {
    name: { type: String, required: true },
    phone: { type: String, required: true },
    licenseNumber: { type: String, required: true, unique: true },
    salaryPerDay: { type: Number, default: 0 },
    salaryPerTrip: { type: Number, default: 0 },
    bataRate: { type: Number, default: 0 },
    totalWorkingHours: { type: Number, default: 0 },
    totalWorkingDays: { type: Number, default: 0 },
    totalTripsCompleted: { type: Number, default: 0 },
    totalBataEarned: { type: Number, default: 0 },
    userId: { type: mongoose.Schema.Types.ObjectId, ref: 'User', unique: true, sparse: true },
    leaves: [leaveSchema],
  },
  { timestamps: true }
);

module.exports = mongoose.model('Driver', driverSchema);
